lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -726,14 +794,35 @@ public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOExcepti
       DocIdSet docIdSet;
       try {
         docIdSet = get(in.getQuery(), cacheHelper);
+      } catch (AsyncQueryLoadInProgressException e) {
+        // Query is being loaded asynchronously. Use uncached version but
+        // do not do a new load of the same query into the cache
+        return in.scorerSupplier(context);
       } finally {
         lock.unlock();
       }
 
       if (docIdSet == null) {
         if (policy.shouldCache(in.getQuery())) {
-          docIdSet = cache(context);
-          putIfAbsent(in.getQuery(), docIdSet, cacheHelper);
+          // If asynchronous caching is requested, perform the same and return
+          // the uncached iterator
+          if (executor != null) {
+            FutureTask<Void> task = new FutureTask<>(() -> {
+              DocIdSet localDocIdSet = cache(context);
+              putIfAbsent(in.getQuery(), localDocIdSet, cacheHelper);
+
+              //remove the key from inflight -- the key is loaded now
+              inFlightAsyncLoadQueries.remove(in.getQuery());
+              return null;
+            });
+            inFlightAsyncLoadQueries.add(in.getQuery());
```

Comment:
I think we have a concurrency bug here -- if the same query across different threads is checked for caching, then two threads can get to this point, and both can compute the cached result and store it?  Because, this `.add` happens at a different time from the above check (`.contains`)...

Maybe this doesn't matter?  It just means two threads do the work of caching, and one is wasted?

Reply:
Yes, that is precisely the idea. The alternative to that is a more aggressive locking which can potentially have a larger negative effect than a thread's caching effort. Thanks for raising this, I have added docs to explain this now.

Reply:
Yeah, that is the idea, since that saves us a bit of pessimism when locking. Thanks for highlighting, I have added docs around the same.